Initial cut of Matter camera app

pidarped opened the pull request:
Basic structure of app with .zap and .matter config.
Chime server added for cluster servers.

#### Testing

pidarped committed:
Initial structure of camera-app

pidarped committed:
Add camera-app.zap and .matter files.

Add chime server support.

Add Chime cluster in the Camera endpoint in .zap config.
Add delegate implementation of chime cluster.

lboue commented on examples/camera-app/linux/args.gni:
```diff
@@ -0,0 +1,27 @@
+chip_project_config_include = "<CHIPProjectAppConfig.h>"
+chip_system_project_config_include = "<SystemProjectConfig.h>"
+chip_project_config_include_dirs =
+    [ "${chip_root}/examples/refrigerator-app/linux/include" ]
```
```suggestion
    [ "${chip_root}/examples/camera-app/linux/include" ]
```

yufengwangca commented on examples/camera-app/camera-common/BUILD.gn:
```diff
@@ -0,0 +1,25 @@
+chip_data_model("camera-common") {
+  zap_file = "camera-app.zap"
+  is_server = true
+}
```
This folder contains the platform independent portion of the app, we need to define a source_set "camera-lib" to build camera-app.h/cpp here instead of each platform

pidarped replied:
Created a source-set for the camera-app files and headers.

yufengwangca commented on examples/camera-app/camera-common/src/camera-app.cpp:
```diff
@@ -0,0 +1,49 @@
+
+void CameraApp::InitCameraDeviceClusters()
+{
+
```
extra  line

pidarped replied:
Fixed.

yufengwangca commented on examples/camera-app/linux/BUILD.gn:
```diff
@@ -0,0 +1,62 @@
+
+executable("camera-app") {
+  sources = [
+    "${chip_root}/examples/camera-app/camera-common/src/camera-app.cpp",
```
camera-app.cpp needed to be moved to camera-common

yufengwangca commented on examples/camera-app/linux/BUILD.gn:
```diff
@@ -0,0 +1,62 @@
+}
+
+group("default") {
+  deps = [ ":linux" ]
```
why default group depends on common group linux, not "camera-app", is it not enough?

pidarped replied:
Changed it to camera-app. But it was primarily to set the linux group as the default target to build.

bzbarsky-apple commented on examples/camera-app/linux/src/clusters/chime/chime-manager.cpp:
```diff
@@ -0,0 +1,68 @@
+    }
+    chimeID = mChimeSounds[chimeIndex].chimeID;
+
+    return chip::CopyCharSpanToMutableCharSpan(mChimeSounds[chimeIndex].name, name);
```
```suggestion
    return CopyCharSpanToMutableCharSpan(mChimeSounds[chimeIndex].name, name);
```
and similar elsewhere.

pidarped committed:
Fix missing device type id in .zap and .matter files for camera device endpoint.

pidarped committed:
Create a source set for the common camera app files.

Also,
- Add build support into gn_build.sh.
- Change executable name to chip-camera-app.

pidarped commented on examples/camera-app/linux/include/camera-device.h:
```diff
@@ -0,0 +1,38 @@
+
+private:
+    // Various cluster server delegates
+    ChimeManager mChimeManager;
```
It is implementing the delegate interface of the cluster but may also have additional functions of interacting with the other components for acting on a command, for example. Furthermore, we have a virtual Delegate interface class which is named as the Delegate.

pidarped committed:
Add camera.md in docs
